### DISCUSSION TITLE
Trying to build a queue for Next js 13 with custom api ( bull, next.js, redis)

### DISCUSSION THREAD

**Discussion by Redbeardjunior** [OptimalBits/bull#2670] [General]
Hi developers i am trying to work on queue in next.js 13 with a custom api.

This is how far I got... the api is working the connection to redis is working but the middle is not working im trying to build the queue to learn and to grow for my personal project.

I'm stuck for 3 days now I did 1 script 2 script and a 3 script use that I just cut it in pieces, still no luck so I try my luck here.

im getting the following error:

```
Error: No .lua files found!
    at loadScripts (webpack-internal:///(rsc)/./node_modules/bull/lib/commands/index.js:29:15)
    at async eval (webpack-internal:///(rsc)/./node_modules/bull/lib/commands/index.js:18:19)
✅ Redis Connected Successfully
api/temp/route.js

```

```
import { NextResponse } from "next/server";
const Queue = require('bull');
import { connectToRedis } from "@/lib/redis";

export async function createQueue() {
  const redis = await connectToRedis();
  const queue = new Queue('my-queue', { redis: { client: redis, }, });
  return queue;
}

export async function addJob(queue, job) {
  await queue.add(job);
}

```
```
export async function POST(request) {
  try {
    const queue = await createQueue();
    await addJob(queue, { name: "John", age: 30 });
    return NextResponse.json({ message: "Hello, redis, bull API" }, { status: 200 });
  } catch (error) {
    console.log(error);
    return NextResponse.json({ message: "An error occured by creating a que."}, { status: 500 });
  }
}

```

second try:

```
const Queue = require('bull');
import { connectToRedis } from "@/lib/redis";

export async function createQueue() {
  const redis = await connectToRedis();
  const queue = new Queue('my-queue', { redis: { client: redis, }, });
  return queue;
}

export async function addJob(queue, job) {
  await queue.add(job);
}
```

```
import { NextResponse } from "next/server";
import Queue from "bull";
import { createQueue, addJob } from "@/lib/queue";

export async function POST(request) {
  try {
    const queue = await createQueue();
    await addJob(queue, { name: "John", age: 30 });
    return NextResponse.json({ message: "Hello, redis, bull API" }, { status: 200 });
  } catch (error) {
    console.log(error);
    return NextResponse.json({ message: "An error occured by creating a que."}, { status: 500 });
  }
}
```
I tried google the awnser, read all the website off the used software but still no luck at all.

**Comment by manast** [comment]
You should use [BullMQ](https://github.com/taskforcesh/bullmq) instead.

**Comment by deadcoder0904** [comment]
@Redbeardjunior did you find a solution? if so, can you post it?

**Reply by Redbeardjunior** [reply]
Goodmorning, I will get back to you !

**Reply by deadcoder0904** [reply]
@Redbeardjunior thanks, i'll try implementing it in the meantime as [this video explained the why behind queues](https://www.youtube.com/watch?v=yXPKB7fHxfo) so i think i can implement it.

plus found 2 other resources:

1. https://medium.com/@asanka_l/integrating-bullmq-with-nextjs-typescript-f41cca347ef8
2. https://dev.to/kishansheth/nextjs-14-booking-app-with-live-data-scraping-using-scraping-browser-610

but do post if you've done something differently. i'll probably implement bullmq with next.js tomorrow for sending emails as the provider i'm using (resend) was unreliable.

**Reply by Redbeardjunior** [reply]
That are some nice steps but not totaly how I fixed it I wil try find time to share some code to get you started. 

because you have to inject some code in your packet.json to make a worker and start that next to your next.js

**Reply by 2019011821** [reply]
waiting

**Reply by pushkarsingh32** [reply]
> That are some nice steps but not totaly how I fixed it I wil try find time to share some code to get you started.
> 
> because you have to inject some code in your packet.json to make a worker and start that next to your next.js

Did you manage to solve it ?

**Comment by Kamleshpaul** [comment]
If you're trying to build a queue in Next.js 13 with Bull and Redis, you might want to consider using BullMQ instead of Bull, as it offers better features and performance.

I faced similar issues when I was setting up my own task queue. If you're still stuck, I wrote a guide that details how to set up queue jobs in Next.js using BullMQ. It includes practical steps and solutions for common errors like the one you're encountering.

You can check it out here: [How to Set Up Queue Jobs in Next.js Using BullMQ](https://codingtricks.co/how-to-set-up-queue-jobs-in-nextjs-using-bullmq).

Feel free to copy and adapt the code as needed!

**Reply by aleksa-codes** [reply]
What about accessing db from the worker? like saving to db? I am having difficulties figuring out how to access db (using drizzle orm) inside a worker.
